lib/web/terminal.go

```diff
@@ -1068,6 +998,199 @@ func (t *WSStream) writeError(err error) error {
 	return trace.Wrap(writeErr)
 }
 
+func (t *WSStream) processMessages() {
+	defer func() {
+		t.cancel()
+		close(t.completedC)
+	}()
+	t.ws.SetReadLimit(teleport.MaxHTTPRequestSize)
+
+	for {
+		select {
+		case <-t.ctx.Done():
+			return
+		default:
+			ty, bytes, err := t.ws.ReadMessage()
+			if err != nil {
+				if errors.Is(err, io.EOF) || errors.Is(err, net.ErrClosed) ||
+					websocket.IsCloseError(err, websocket.CloseAbnormalClosure, websocket.CloseGoingAway, websocket.CloseNormalClosure) {
+					return
+				}
+
+				msg := err.Error()
+				if len(bytes) > 0 {
+					msg = string(bytes)
+				}
+				select {
+				case <-t.ctx.Done():
+					return
+				case t.rawC <- Envelope{Payload: msg}:
+					return
+				}
+			}
+
+			if ty != websocket.BinaryMessage {
+				t.log.Warnf("Expected binary message, got %v", ty)
```

Comment:
At that point, we should probably return an error to the client. Without it client potentially will wait for a response that will never come.